firebase-firestore/src/main/java/com/google/firebase/firestore/local/IndexedQueryEngine.java

```diff
@@ -69,18 +71,64 @@ public void setIndexManager(IndexManager indexManager) {
     hardAssert(indexManager != null, "setIndexManager() not called");
 
     // Queries that match all documents don't benefit from index-based lookups.
-    if (query.matchesAllDocuments()) {
+    // TODO(ehsann): remove `query.containsCompositeFilters()` from this condition to serve
+    // composite queries from the index.
+    if (query.matchesAllDocuments() || query.containsCompositeFilters()) {
       return executeFullCollectionScan(query);
     }
 
     Target target = query.toTarget();
 
-    FieldIndex fieldIndex = indexManager.getFieldIndex(target);
-    // TODO(indexing): Fall back to DefaultQueryEngine if index-free queries is available
+    if (target.getDnf().size() == 0) {
+      // There are no filters.
+      return performCollectionQueryForFilter(query, target, null);
+    }
+
+    ImmutableSortedMap<DocumentKey, Document> result =
+        ImmutableSortedMap.Builder.emptyMap(DocumentKey.comparator());
+    for (CompositeFilter andFilter : target.getDnf()) {
+      // Each filter in the DNF must be an AND filter.
+      hardAssert(andFilter.isAnd(), "Found an OR filter in the DNF");
+      for (Map.Entry<DocumentKey, Document> entry :
+          performCollectionQueryForFilter(query, target, andFilter)) {
+        result = result.insert(entry.getKey(), entry.getValue());
+      }
+    }
+
+    // If there's no limit constraint, we can return all the results.
+    // If the DNF contained only 1 term, the SQLite query has enforced the limit, and we can return
+    // all the results.
+    if (target.getLimit() == -1 || target.getDnf().size() == 1) {
+      return result;
+    }
+
+    // If there's a limit constraint, we should perform all branches of the query (as done above),
```

Comment:
The limit handling should be pushed into SQLite.

Reply:
Acknowledged.